### ISSUE TITLE
Adding GPU information

### ISSUE THREAD

**Issue by dbwiddis** [oshi/oshi#2112]
GPU information has been requested in #17, #400, and #1248 so there is definitely interest in GPU information.   While Windows and macOS have APIs for some information, the Linux tools for this are diverse (at least 3 major hardware vendors) and essentially require parsing "pretty" command-line output.  This makes it very difficult to develop a consistent cross-platform API.

All that said, I'm going to open this issue to describe how to implement such a capability, and will tag it with [up-for-grabs](https://up-for-grabs.net/#/) to offer anyone who wants to do so to follow the steps below to add this feature.

### Decide where to put the new method(s).

#17 and #400 primarily requested graphics card information.  This has been implemented in the [GraphicsCard](https://www.oshi.ooo/oshi-core-java11/apidocs/com.github.oshi/oshi/hardware/GraphicsCard.html) class which includes the video ram values.   This could be a potential location to add GPU usage information, but may not be the best as OSHI returns a list of these and as far as I'm aware the "Usage %" value available is a single percentage.  

Sometimes GPUs are integrated into the CPU.  And "% GPU Usage" is a natural parallel of "% CPU Usage" on the Processor object.  But that is inaccurate for separate graphics cards.  So maybe we need both, or a systemwide number one place and a per-card number another place?  This is a puzzle that whoever implements this will need to solve.

### Add the new method(s) at the interface level.

Add `getGpuUsage()` (or whatever methods you choose to implement) on the appropriate interface you determined in the previous step.   Add javadocs explaining what information is expected and what platforms it's implemented on.  Use an appropriate `default` return value.

### Implement on at least one operating system.

On the appropriate platform-specific implementation of the interface in the previous step, write the code that obtains the values.

You may find as you do this that your envisioned interface should be different.  Feel free to iteratively change the interface.  

This may be a good point to submit a PR to OSHI with a draft of what you have so far for feedback.

### Implement on another operating system.

Do a second platform-specific implementation.  You might find that you need more or fewer methods or a different more general cross-platform name for the value you're returning.

### Implement on remaining operating systems and hardware that you have access to debug and test.

For this task in particular, for the Linux implementation you probably only have one GPU vendor's software on your machine.  Just implement that and leave it for others to add their own implementations later.  Return a sensible default and perhaps a useful log message.

### Once you've done all you can, change your PR from Draft to Ready for Review.  

Once we're happy with it, we'll merge it and request others contribute additional OS or GPU Hardware code.   Ideally this will occur in between releases and we can continue to change the API if required.  Once we cut a release, we're stuck with the existing API (we can add to it, but not remove anything).  

### macOS

This is probably the easiest implementation and I'm even happy to do this part once you've done another OS.

Basic code logic to retrieve it is on [Stack Overflow](https://stackoverflow.com/questions/10110658/programmatically-get-gpu-percent-usage-in-os-x). Note in the comments that some names have changed. You will have to port that to Java/JNA but there are many examples in OSHI of navigating the registry such as [here](https://github.com/oshi/oshi/blob/356942e1974ba7d7cd824f64175ca06001411928/oshi-core/src/main/java/oshi/hardware/platform/mac/MacDisplay.java#L71-L97).   There are registry entries for each graphics card, so this actually may be a better implementation to put on the `GraphicsCard` object.

The `PerformanceInfo` dictionary property has 76 values (Radeon) / 82 values (integrated Intel).  Useful ones on my machine that differ from the SO answer:
 - Device Utilization % (Integrated card only, and not sure how to interpret the value.  It seems a constant "3" on my system.)
 - inUseSysMemoryBytes
 - inUseVidMemoryBytes
 - gartFreeBytes
 - gartUsedBytes
 - gartSizeBytes
Looks like a broad collection of possible outputs [is collected here](https://gist.github.com/chockenberry/2afe4d0f1f9caddc81de?permalink_comment_id=1615189).

### Windows

It looks like there are solutions out there using [`D3DKMTQueryStatistics`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/d3dkmthk/nf-d3dkmthk-d3dkmtquerystatistics) but the documentation says:
> Reserved for system use. Do not use in your driver.

That said, looks like lots of people are using this.  [This SO answer](https://stackoverflow.com/questions/29611611/c-windows-gpu-usage-load) has some links, but of note there is no one "usage" number; one has to measure an overall usage value and divide by a time delta to get a usage percentage (not unlike CPU ticks).   So this doesn't look like a good approach.

However.... based on [this SO answer](https://stackoverflow.com/questions/56830434/c-sharp-get-total-usage-of-gpu-in-percentage) it looks like there are GPU usage performance counters.  I don't see them on my Win 7 VM, but someone with newer Windows version could probably test OSHI's PerfMon (wildcard) counters for "GPU Engine", and a "Utilization Percentage" counter.  It doesn't appear that there's a WMI table exposing these counters to use as a backup, though.

It looks like NVidia implements a namespace for WMI: see https://documentation.help/nvWmi/documentation.pdf .

### Linux

There appears to be command-line parsing of several major card vendors.  I wouldn't even attempt to do this without having the appropriate hardware in a machine to print the output.  Perhaps users of such hardware could add comments below with the textual output to at least allow development.

Here's a summary.  Seems "usage" is a memory used/total thing in many outputs, though.  https://unix.stackexchange.com/questions/38560/gpu-usage-monitoring-cuda

### Unix

I suspect some overlap with Linux here.  Haven't even looked at it.

**Comment by harmonjeff** [comment]
I am a total beginner at this.  As this is my first attempt, the code sample might not be the most efficient way to get the information.  This is working on my 14" M1 MacBook Pro (MacBookPro18,3) for GPU Processor Utilization, GPU Used Memory, GPU Total Memory.  Might help as a starting point for something that works across more Macs.

```
private final static String GPU_CORE_USE_KEY = "Device Utilization %";
  private final static String GPU_TOTAL_VRAM_KEY = "Alloc system memory";
  private final static String GPU_USED_VRAM_KEY = "In use system memory";

  public static GraphicsUsage getGraphicsUsage() throws UnsupportedGraphicsException {
    int coreUse = 0;
    long totalVram = 0;
    long usedVram = 0;
    IORegistryEntry ioAccelerator = IOKitUtil.getMatchingService("IOAccelerator");
    if (ioAccelerator != null) {
      LOG.debug("IOAcceleratorName: {}", ioAccelerator.getName());
      if (!"AGXAcceleratorG13X".equals(ioAccelerator.getName())) {
        LOG.error("Can't get metrics for {} graphics",ioAccelerator.getName());
        throw new UnsupportedGraphicsException(
            String.format("%s is not yet supported for gathering metrics", ioAccelerator.getName()));
      }

      CFStringRef performanceStatisticsKey = CFStringRef.createCFString("PerformanceStatistics");
      CFTypeRef performanceStatisticsRaw = ioAccelerator.createCFProperty(performanceStatisticsKey);
      if (performanceStatisticsRaw != null) {
        CFMutableDictionaryRef performanceStatisticsRef = new CFMutableDictionaryRef(
            performanceStatisticsRaw.getPointer());

        CFStringRef key = CFStringRef.createCFString(GPU_CORE_USE_KEY);
        CFNumberRef stat = new CFNumberRef(performanceStatisticsRef.getValue(key));
        coreUse = stat.intValue();
        stat.release();
        key.release();

        key = CFStringRef.createCFString(GPU_TOTAL_VRAM_KEY);
        stat = new CFNumberRef(performanceStatisticsRef.getValue(key));
        totalVram = stat.longValue();
        stat.release();
        key.release();

        key = CFStringRef.createCFString(GPU_USED_VRAM_KEY);
        stat = new CFNumberRef(performanceStatisticsRef.getValue(key));
        usedVram = stat.longValue();
        stat.release();
        key.release();

        performanceStatisticsRef.release();
        performanceStatisticsRaw.release();
      }
      performanceStatisticsKey.release();
      ioAccelerator.release();
    }
    return GraphicsUsage.builder()
        .coreUsage(coreUse)
        .totalVramBytes(totalVram)
        .usedVramBytes(usedVram)
        .build();
  }
```

**Comment by harmonjeff** [comment]
Hmm, have had my Azul 17 JVM (Zulu17.34+19-CA (17.0.3+7) (build 17.0.3+7-LTS)) crash a few times with this code running.  Not immediately, sometimes it has gone as long as 20 minutes, but multiple JVM crashes.

Just sharing information that may help in the development here.

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000001b65d49c0, pid=39451, tid=44547
#
# JRE version: OpenJDK Runtime Environment Zulu17.34+19-CA (17.0.3+7) (build 17.0.3+7-LTS)
# Java VM: OpenJDK 64-Bit Server VM Zulu17.34+19-CA (17.0.3+7-LTS, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)
# Problematic frame:
# C  [libobjc.A.dylib+0x99c0]  objc_release+0x10
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# If you would like to submit a bug report, please visit:
#   http://www.azul.com/support/
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

**Comment by dbwiddis** [comment]
I'm thinking you are double-releasing this pointer:
```
CFMutableDictionaryRef performanceStatisticsRef = new CFMutableDictionaryRef(performanceStatisticsRaw.getPointer());
```
This essentially makes `performanceStatisticsRef` and `performanceStatisticsRaw` the same pointer.  Then later:
```
performanceStatisticsRef.release();
performanceStatisticsRaw.release();
```
You only have to release things you "create" and you didn't create the Ref pointer so delete that line.

As for the code, looks reasonable at first glance.  But I mentioned macOS is the easy one.  Also note that the dictionary keys have changed over time, [see here](https://gist.github.com/chockenberry/2afe4d0f1f9caddc81de?permalink_comment_id=1615189).  So you might need to handle more than one potential key.

**Comment by harmonjeff** [comment]
@dbwiddis thank you!  I bet you are right.  I'll take that out and give it a try.

**Comment by harmonjeff** [comment]
Sure enough.  Had it running and gather GPU metrics for over an hour without issue after getting rid of that bad release.  Thank you!

**Comment by dbwiddis** [comment]
Other helpful advice: if you're going to do this in a loop you might want to do all the CFString creation outside the loop.

**Comment by dbwiddis** [comment]
## GPU Usage Metrics — Implementation Plan

### Overview

Add dynamic GPU usage metrics to OSHI, complementing the existing static `GraphicsCard` info. The approach follows two phases: Phase 1 uses only native OS APIs that OSHI already knows how to talk to (PDH, sysfs, IOKit). Phase 2 adds optional vendor-specific libraries loaded at runtime for richer data.

---

### Phase 1 — Cross-Platform, No Vendor Libraries

| Metric | Windows (PDH) | Linux (sysfs) | macOS (IOKit) |
|---|---|---|---|
| **GPU Utilization (ticks)** | Cumulative Running Time via `\GPU Engine(*)\Running Time` (Win10 1709+) | Not available | Not available |
| **GPU Utilization (instant %)** | Not available (use ticks) | `gpu_busy_percent` (AMD amdgpu) / freq ratio proxy (Intel i915) | Scaled integer via `IOAccelerator` PerformanceStatistics |
| **VRAM Used** | `\GPU Adapter Memory(*)\Dedicated Usage` | `mem_info_vram_used` (AMD amdgpu) | `vramUsedBytes` from IOKit |
| **Shared Memory Used** | `\GPU Adapter Memory(*)\Shared Usage` | N/A | N/A |

**Key design decisions:**

- **Atomic snapshots.** GPU ticks and clock timestamp are captured in a single call and returned as one immutable object. There are no separate getters that could introduce skew between the timestamp and the tick value.
- **Two access patterns for utilization.** Windows provides native cumulative tick counters (like CPU ticks) via `getGpuTicks()`. Linux and macOS provide instantaneous percentages via `getGpuUtilization()`. The caller checks which is available and uses the appropriate method.
- **Windows tick aggregation.** Windows PDH exposes per-engine running times (3D, Copy, VideoDecode, VideoEncode, Compute). Since GPU engines are parallel pipelines, overall utilization is the max across engines, not the sum. This aggregation is done internally; the caller sees a single active tick value.
- **Intel on Linux is limited in Phase 1.** The i915 driver exposes utilization through `perf_event` PMU counters which require the perf syscall, not simple file reads. A frequency ratio (`rps_act_freq_mhz / rps_max_freq_mhz`) serves as a rough proxy. Full Intel support improves in Phase 2.
- **NVIDIA on Linux requires Phase 2.** The proprietary driver exposes nothing useful in sysfs. All NVIDIA metrics require NVML.

---

### Phase 2 — Vendor Libraries, Temperature, and Richer Metrics

Loaded optionally at runtime via JNA. If the library is not present, graceful fallback to Phase 1 sources.

| Source | Coverage | Additional Metrics |
|---|---|---|
| **NVML** (`libnvidia-ml.so` / `nvml.dll`) | NVIDIA on Windows + Linux | Utilization, VRAM, temperature, power draw, clocks, fan speed, encoder/decoder utilization |
| **ADL / ROCm-SMI** | AMD on Windows + Linux | Utilization, temperature, power, clocks, fan speed |
| **Linux hwmon** | AMD + Intel | Temperature (requires correlating hwmon device to DRM card) |

**Temperature is deferred to Phase 2.** macOS Apple Silicon GPU temps require private IOReport APIs that Apple keeps changing. Windows `MSAcpi_ThermalZoneTemperature` is unreliable for GPU-specific readings. Linux hwmon works but needs non-trivial device correlation logic. All three platforms get reliable GPU temperature through vendor libraries (NVML, ADL) or with additional hwmon mapping work that pairs well with the vendor library integration effort.

---

### Draft Interfaces

```java
/**
 * Atomic snapshot of GPU active time and wall clock timestamp.
 * Captured in a single call to ensure no skew between ticks and time.
 *
 * Call {@link GraphicsCard#getGpuTicks()} twice and compute deltas
 * between snapshots to derive utilization:
 *
 *   utilization% = (active2 - active1) / (timestamp2 - timestamp1) * 100
 */
public interface GpuTicks {
    /**
     * Wall clock time of this snapshot in 100-nanosecond units,
     * same scale as the active tick value.
     *
     * @return wall clock timestamp in 100ns units
     */
    long getTimestamp();

    /**
     * Cumulative GPU active time in 100-nanosecond units.
     * Monotonically increasing.
     *
     * On platforms with per-engine counters (Windows), this is the
     * maximum across all engine types, since GPU engines are parallel
     * pipelines.
     *
     * @return cumulative active ticks, or 0 if tick-level metrics
     *         are not available on this platform
     */
    long getActiveTicks();
}
```
```java
public interface GraphicsCard {
    // ... existing static methods (getName, getDeviceId, getVendor,
    //     getVersionInfo, getVRam) unchanged ...

    /**
     * Atomic snapshot of cumulative GPU active time and wall clock
     * timestamp. Call twice and compute deltas to derive utilization.
     *
     * Primarily available on Windows 10 1709+ via PDH Running Time
     * counters. On platforms without tick-level support, returns a
     * snapshot with 0 activeTicks.
     *
     * @return an immutable GpuTicks snapshot, never null
     */
    GpuTicks getGpuTicks();

    /**
     * Instantaneous GPU utilization as a percentage (0.0 - 100.0).
     *
     * Available on Linux (AMD gpu_busy_percent, Intel freq ratio)
     * and macOS (IOKit GPU Core Utilization).
     *
     * @return utilization percentage, or -1 if not available
     */
    double getGpuUtilization();

    /**
     * VRAM currently in use, in bytes.
     *
     * @return bytes of VRAM in use, or -1 if unavailable
     */
    long getVramUsed();

    /**
     * Shared system memory currently used by this GPU, in bytes.
     * Primarily available on Windows (WDDM shared memory).
     *
     * @return bytes of shared memory in use, or -1 if unavailable
     */
    long getSharedMemoryUsed();
}
```
Caller Usage Pattern
```java
GraphicsCard gpu = ...;

// Prefer ticks if available (Windows), fall back to instant percentage
GpuTicks snapshot = gpu.getGpuTicks();
if (snapshot.getActiveTicks() > 0) {
    // Store snapshot, compute delta on next poll:
    //   util% = (current.getActiveTicks() - previous.getActiveTicks())
    //         / (current.getTimestamp() - previous.getTimestamp()) * 100
} else {
    double util = gpu.getGpuUtilization();
    if (util >= 0) {
        // Use directly
    }
}

long vramUsed = gpu.getVramUsed();
long sharedUsed = gpu.getSharedMemoryUsed();
```
